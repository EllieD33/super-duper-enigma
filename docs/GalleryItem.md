# GalleryItem Component

## Description

The `GalleryItem` component renders an image inside a container, with optional overlay text. It supports different sizes and displays a loading skeleton or fallback image in case of errors or while loading.

## Props

### `GalleryItemProps`

| Prop             | Type                             | Required | Description                                                                                               |
| ---------------- | -------------------------------- | -------- | --------------------------------------------------------------------------------------------------------- |
| `imageUrl`       | `string`                         | ✅       | The URL of the image to display.                                                                          |
| `altText`        | `string`                         | ✅       | The alt text for the image.                                                                               |
| `imageId`        | `number`                         | ✅       | Unique identifier for the image.                                                                          |
| `size`           | `"small" \| "medium" \| "large"` | ❌       | Optional size of the image. Defaults to `"medium"`.                                                       |
| `overlayText`    | `string`                         | ❌       | Optional text to display over the image as an overlay.                                                    |
| `overlayDisplay` | `"hover" \| "static"`            | ❌       | Specifies when the overlay text should appear: `"hover"` for hover effect, `"static"` for always visible. |

## Behaviour

- The `GalleryItem` component renders an image inside a container.
- If the image is still loading, a skeleton loader will be displayed.
- If the image fails to load, a fallback image will be shown instead.
- Optional overlay text can be displayed, either always visible (`"static"`) or shown only on hover (`"hover"`).
- Supports multiple sizes of images: small, medium, and large.
- The component uses a lazy loading strategy for images to improve performance.

[Explore GalleryItem in Storybook](http://localhost:6006/?path=/docs/library-galleries-galleryitem--docs&args=error:&globals=viewport:largeMobile)

## Notes

- **Fallback Image**: If the `imageUrl` fails to load or has an error, a placeholder image (`https://placehold.co/600`) is shown instead.
- **Overlay Display**: When `overlayDisplay` is set to `"hover"`, the overlay text appears only when the user hovers over the image. If set to `"static"`, the text remains visible at all times.
- The **loading skeleton** is shown while the image is still loading, providing a better user experience during image load times.
- The `GalleryItem` component has a flexible design, adapting to different image sizes depending on the `size` prop.
