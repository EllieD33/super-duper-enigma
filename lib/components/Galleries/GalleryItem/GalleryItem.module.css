.image {
    max-width: 100%;
    height: auto;
    border-radius: 15px;
    object-fit: cover;
}