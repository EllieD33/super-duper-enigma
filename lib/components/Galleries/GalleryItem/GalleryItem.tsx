import React, { ReactElement } from "react";
import styles from "./GalleryItem.module.css";
import clsx from "clsx";

export interface GalleryItemProps {
    imageUrl: string;
    altText: string;
    imageId: number;
    size?: "small" | "medium" | "large";
}

const GalleryItem = ({
    imageUrl,
    altText,
    imageId,
    size = "medium",
}: GalleryItemProps): ReactElement => {
    const galleryItemStyles = clsx(styles.image, styles[size]);

    return (
        <div id={`gallery-item-${imageId}`}>
            <img
                className={galleryItemStyles}
                src={imageUrl}
                alt={altText}
                loading="lazy"
            />
        </div>
    );
};

export default GalleryItem;
