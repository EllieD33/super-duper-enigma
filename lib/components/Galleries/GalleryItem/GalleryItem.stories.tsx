import type { Meta, StoryObj } from "@storybook/react";
import GalleryItem, { GalleryItemProps } from "./GalleryItem";

const meta = {
    title: "Library/Galleries/GalleryItem",
    component: GalleryItem,
    parameters: {
        layout: "centered",
    },
    tags: ["autodocs"],
    argTypes: {},
} satisfies Meta<typeof GalleryItem>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
    args: {
        imageId: 1,
        imageUrl: "https://cdn.pixabay.com/photo/2023/08/08/15/01/flower-8177578_640.jpg",
        altText: "white flowers wth dark green leaves behind them",
    } as GalleryItemProps,
};
