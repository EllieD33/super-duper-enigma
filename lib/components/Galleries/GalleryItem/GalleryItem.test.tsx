import React from "react";
import { render } from "@testing-library/react";
import GalleryItem, { GalleryItemProps } from "./GalleryItem";

const defaultProps: GalleryItemProps = {
    imageId: 1,
    imageUrl: "https://example.com/sunset.jpg",
    altText: "beautiful sunset",
};

describe("GalleryItem", () => {
    it("should render the component medium size by default", () => {
        const { container, getByAltText } = render(
            <GalleryItem {...defaultProps} />
        );
        const image = getByAltText(defaultProps.altText);

        expect(image).toBeInTheDocument();
        expect(image).toHaveClass("medium");
        expect(container).toMatchSnapshot();
    });

    it("should match snapshot for all variants", () => {
        const sizes = ["small", "medium", "large"] as const;
        sizes.forEach((size) => {
            const { container } = render(
                <GalleryItem {...defaultProps} size={size} />
            );
            expect(container).toMatchSnapshot();
        });
    });

    it("should have an alt attribute matching the altText prop", () => {
        const { getByAltText } = render(<GalleryItem {...defaultProps} />);
        const image = getByAltText(defaultProps.altText);

        expect(image).toHaveAttribute("alt", defaultProps.altText);
    });
});
