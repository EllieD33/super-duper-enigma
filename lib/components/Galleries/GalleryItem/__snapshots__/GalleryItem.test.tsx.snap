// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GalleryItem should match snapshot for all variants 1`] = `
<div>
  <div
    id="gallery-item-1"
  >
    <img
      alt="beautiful sunset"
      class="image small"
      loading="lazy"
      src="https://example.com/sunset.jpg"
    />
  </div>
</div>
`;

exports[`GalleryItem should match snapshot for all variants 2`] = `
<div>
  <div
    id="gallery-item-1"
  >
    <img
      alt="beautiful sunset"
      class="image medium"
      loading="lazy"
      src="https://example.com/sunset.jpg"
    />
  </div>
</div>
`;

exports[`GalleryItem should match snapshot for all variants 3`] = `
<div>
  <div
    id="gallery-item-1"
  >
    <img
      alt="beautiful sunset"
      class="image large"
      loading="lazy"
      src="https://example.com/sunset.jpg"
    />
  </div>
</div>
`;

exports[`GalleryItem should render the component medium size by default 1`] = `
<div>
  <div
    id="gallery-item-1"
  >
    <img
      alt="beautiful sunset"
      class="image medium"
      loading="lazy"
      src="https://example.com/sunset.jpg"
    />
  </div>
</div>
`;
